src/SafeAreaContext.tsx

```diff
@@ -31,78 +31,101 @@ export interface SafeAreaProviderProps extends ViewProps {
+      initialMetrics?.frame ??
+        parentFrame ?? {
+          // Backwards compat so we render anyway if we don't have frame.
+          x: 0,
+          y: 0,
+          width: Dimensions.get('window').width,
+          height: Dimensions.get('window').height,
+        },
+    );
+    const onInsetsChange = React.useCallback(
+      (event: InsetChangedEvent) => {
+        const {
+          nativeEvent: { frame: nextFrame, insets: nextInsets },
+        } = event;
+
+        setFrame((frame) => {
+          if (
+            // Backwards compat with old native code that won't send frame.
+            nextFrame &&
+            (nextFrame.height !== frame.height ||
+              nextFrame.width !== frame.width ||
+              nextFrame.x !== frame.x ||
+              nextFrame.y !== frame.y)
+          ) {
+            return nextFrame;
+          } else {
+            return frame;
+          }
+        });
+
+        setInsets((insets) => {
+          if (
+            !insets ||
+            nextInsets.bottom !== insets.bottom ||
+            nextInsets.left !== insets.left ||
+            nextInsets.right !== insets.right ||
+            nextInsets.top !== insets.top
+          ) {
+            return nextInsets;
+          } else {
+            return insets;
+          }
+        });
       },
-  );
-  const onInsetsChange = React.useCallback(
-    (event: InsetChangedEvent) => {
-      const {
-        nativeEvent: { frame: nextFrame, insets: nextInsets },
-      } = event;
-
-      if (
-        // Backwards compat with old native code that won't send frame.
-        nextFrame &&
-        (nextFrame.height !== frame.height ||
-          nextFrame.width !== frame.width ||
-          nextFrame.x !== frame.x ||
-          nextFrame.y !== frame.y)
-      ) {
-        setFrame(nextFrame);
+      [setFrame, setInsets],
+    );
+
+    // Add flex to the style if it doesn't have it.
+    const flexedStyle = React.useMemo(() => {
+      const flatStyle = StyleSheet.flatten(style) || {};
+      if ('flex' in flatStyle === false) {
+        return {
+          ...flatStyle,
+          flex: 1,
+        };
       }
 
-      if (
-        !insets ||
-        nextInsets.bottom !== insets.bottom ||
-        nextInsets.left !== insets.left ||
-        nextInsets.right !== insets.right ||
-        nextInsets.top !== insets.top
-      ) {
-        setInsets(nextInsets);
-      }
-    },
-    [frame, insets],
-  );
-
-  return (
-    <NativeSafeAreaProvider
-      style={[styles.fill, style]}
-      onInsetsChange={onInsetsChange}
-      {...others}
-    >
-      {insets != null ? (
-        <SafeAreaFrameContext.Provider value={frame}>
-          <SafeAreaInsetsContext.Provider value={insets}>
-            {children}
-          </SafeAreaInsetsContext.Provider>
-        </SafeAreaFrameContext.Provider>
-      ) : null}
-    </NativeSafeAreaProvider>
-  );
-}
+      return style;
+    }, [style]);
```

Comment:
We don't need this - just use `StyleSheet.compose(styles.fill, style)`

Reply:
Oh cool, I didn't know of that util. Will change this. Thanks for the fast review.

Reply:
Pushed up `StyleSheet.compose`.

Looking at the code for this util, it has the same issue if the dev passes in a custom `style` prop. It will return an array, breaking the referential equality - https://github.com/facebook/react-native/blob/f2447e6048a6b519c3333767d950dbf567149b75/packages/react-native/Libraries/StyleSheet/StyleSheet.js#L275

If we do this, then do we need to worry about referential equality for `onInsetsChange`?

Reply:
The style memoization can end up being more expensive if users pass in inline styles

> If we do this, then do we need to worry about referential equality for onInsetsChange?

With your change, you can now set the dependencies to `[]`, since `setFrame` and `setInsets` are stable references. So I'd say since it's guaranteed to only ever have one value, it's worth it

Reply:
Oh for sure. I'll update this a little later today please. Thanks!

Reply:
Updated! Should I squash everything up into one commit?